File: app/src/main/java/net/syncthing/lite/fragments/FoldersFragment.java
package net.syncthing.lite.fragments;

import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;

import com.google.common.collect.Lists;
import com.google.common.collect.Ordering;

import net.syncthing.lite.R;
import net.syncthing.lite.activities.FolderBrowserActivity;
import net.syncthing.lite.activities.SyncthingActivity;
import net.syncthing.lite.adapters.FoldersListAdapter;
import net.syncthing.lite.databinding.FragmentFoldersBinding;

import org.apache.commons.lang3.tuple.Pair;

import java.util.Collections;
import java.util.List;

import it.anyplace.sync.core.beans.FolderInfo;
import it.anyplace.sync.core.beans.FolderStats;

public class FoldersFragment extends Fragment {

    private static final String TAG = "FoldersFragment";

    private SyncthingActivity mActivity;
    private FragmentFoldersBinding mBinding;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        mBinding = DataBindingUtil.inflate(getLayoutInflater(), R.layout.fragment_folders, container, false);
        mBinding.list.setEmptyView(mBinding.empty);
        return mBinding.getRoot();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        mActivity = (SyncthingActivity) getActivity();
        showAllFoldersListView();
    }

    private void showAllFoldersListView() {
        List<Pair<FolderInfo, FolderStats>> list = Lists.newArrayList(mActivity.getFolderBrowser().getFolderInfoAndStatsList());
        Collections.sort(list, Ordering.natural().onResultOf(input -> input.getLeft().getLabel()));
        Log.i(TAG, "list folders = " + list + " (" + list.size() + " records");
        ArrayAdapter<Pair<FolderInfo, FolderStats>> adapter = new FoldersListAdapter(getContext(), list);
        mBinding.list.setAdapter(adapter);
        mBinding.list.setOnItemClickListener((adapterView, view, position, l) -> {
            String folder = adapter.getItem(position).getLeft().getFolder();
            Intent intent = new Intent(getContext(), FolderBrowserActivity.class);
            intent.putExtra(FolderBrowserActivity.EXTRA_FOLDER_NAME, folder);
            startActivity(intent);
        });
    }
}

File: app/src/main/kotlin/net/syncthing/lite/fragments/FoldersFragment.kt
package net.syncthing.lite.fragments

import android.arch.lifecycle.Observer
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import net.syncthing.java.core.beans.FolderInfo
import net.syncthing.java.core.beans.FolderStats
import net.syncthing.lite.activities.FolderBrowserActivity
import net.syncthing.lite.adapters.FolderListAdapterListener
import net.syncthing.lite.adapters.FoldersListAdapter
import net.syncthing.lite.databinding.FragmentFoldersBinding
import org.jetbrains.anko.intentFor

class FoldersFragment : SyncthingFragment() {

    private val TAG = "FoldersFragment"

    private lateinit var binding: FragmentFoldersBinding

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        binding = FragmentFoldersBinding.inflate(layoutInflater, container, false)

        libraryHandler.isListeningPortTaken.observe(this, Observer { binding.listeningPortTaken = it })

        return binding.root
    }

    override fun onLibraryLoaded() {
        showAllFoldersListView()
    }

    private fun showAllFoldersListView() {
        libraryHandler.folderBrowser { folderBrowser ->
            val list = folderBrowser.folderInfoAndStatsList()

            GlobalScope.launch (Dispatchers.Main) {
                Log.i(TAG, "list folders = " + list + " (" + list.size + " records)")
                val adapter = FoldersListAdapter().apply { data = list }
                binding.list.adapter = adapter
                adapter.listener = object : FolderListAdapterListener {
                    override fun onFolderClicked(folderInfo: FolderInfo, folderStats: FolderStats) {
                        startActivity(
                                activity!!.intentFor<FolderBrowserActivity>(
                                        FolderBrowserActivity.EXTRA_FOLDER_NAME to folderInfo.folderId
                                )
                        )
                    }
                }

                binding.isEmpty = list.isEmpty()
            }
        }
    }

    override fun onIndexUpdateComplete(folderInfo: FolderInfo) {
        super.onIndexUpdateComplete(folderInfo)
        showAllFoldersListView()
    }
}
